### PR TITLE
NodeJS homework week 2 (Nojdar)

### PR DESCRIPTION
**Pull request by NojdarTaho**

### CODE REVIEW

File: homework/hackyourtemperature/app.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+// Create a post route that has /weather as an endpoint:
+app.post("/weather", async (req, res) => {
+  try {
+    const cityName = req.body.cityName;
+    const data = await fetch(
+      `https://api.openweathermap.org/data/2.5/weather?q=${cityName}&appid=${keys.API_KEY}&units=metric`
+    );
+    const jsonData = await data.json();
+    const temperature = jsonData.main.temp;
+    res.send({ City: cityName, Temperature: temperature });
```
Comment (allhandsondeck) [inline homework/hackyourtemperature/app.js:22]: As per naming convention of object properties, please use camel-case for properties.

File: homework/hackyourtemperature/app.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+app.get("/", (req, res) => {
+  res.send("<h1>Hello from backend to frontend!!</h1>");
+});
+
+// Create a post route that has /weather as an endpoint:
+app.post("/weather", async (req, res) => {
+  try {
+    const cityName = req.body.cityName;
+    const data = await fetch(
+      `https://api.openweathermap.org/data/2.5/weather?q=${cityName}&appid=${keys.API_KEY}&units=metric`
```
Comment (allhandsondeck) [inline homework/hackyourtemperature/app.js:18]: Nice!

File: homework/hackyourtemperature/_tests_/app.test.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    });
+    expect(response.header["content-type"]).toEqual(
+      expect.stringContaining("json")
+    );
+  });
+  it("It should respond with temperature of the city", async () => {
+    const response = await request.post("/weather").send({
+      cityName: "London",
+    });
+    expect(typeof response.body["Temperature"]).toEqual("number");
```
Comment (allhandsondeck) [inline homework/hackyourtemperature/_tests_/app.test.js:26]: I think this assertion does not match with the spec on ln22 because you are asserting on type of `Temperature` property not the value of `Temperature` property.
How can you fix this?

File: homework/hackyourtemperature/_tests_/app.test.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    });
+    expect(response.statusCode).toBe(200);
+  });
+
+  it("It should specify json in the content type header", async () => {
+    const response = await request.post("/weather").send({
+      cityName: "London",
+    });
+    expect(response.header["content-type"]).toEqual(
+      expect.stringContaining("json")
```
Comment (allhandsondeck) [inline homework/hackyourtemperature/_tests_/app.test.js:19]: `stringContaining` is a good assertion but in tests we would like to be as precise as possible.
So which other methods of `expect` can we use instead of `stringContaining` to just assert on the value of `content-type`?

File: homework/hackyourtemperature/_tests_/app.test.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+// When things go wrong scenarios
+
+describe("cityName is missing or wrong", () => {
+  it("It should respond with a 400 status code if the cityName is wrong", async () => {
+    const response = await request.post("/weather").send({
+      cityName: "randomString",
+    });
+    expect(response.statusCode).toBe(400);
+  });
+  it("It should respond with a 400 status code if the cityName is missing", async () => {
```
Comment (allhandsondeck) [inline homework/hackyourtemperature/_tests_/app.test.js:44]: Nice that you are testing also edge cases 👍

File: homework/hackyourtemperature/_tests_/app.test.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    expect(typeof response.body["Temperature"]).toEqual("number");
+  });
+  it("It should respond with name of the city", async () => {
+    const response = await request.post("/weather").send({
+      cityName: "London",
+    });
+    expect(response.body["City"]).toEqual("London");
+  });
+});
+// When things go wrong scenarios
```
Comment (allhandsondeck) [inline homework/hackyourtemperature/_tests_/app.test.js:35]: Especially in testing, you don't much need comments because one of the most beneficial aim in testing is providing documentation with the scenarios you are testing. So you are already documenting below with `describe` and `it` that you are testing unhappy scenarios.

File: homework/hackyourtemperature/_tests_/app.test.js (hunk 3 above)
Comment (allhandsondeck) [reply homework/hackyourtemperature/_tests_/app.test.js:26]: Or, you can assert on the existence of the `Temperature` property instead of it's type.

File: homework/hackyourtemperature/jest.config.js [hunk 7]
```diff
@@ -0,0 +1,8 @@
+export default {
+  // Tells jest that any file that has 2 .'s in it and ends with either js or jsx should be run through the babel-jest transformer
+  transform: {
+    "^.+\\.jsx?$": "babel-jest",
+  },
+  // By default our `node_modules` folder is ignored by jest, this tells jest to transform those as well
+  transformIgnorePatterns: [],
```
Comment (allhandsondeck) [inline homework/hackyourtemperature/jest.config.js:7]: Is this property auto-generated?
If it isn't. Please delete it because I don't think it does anything without some items inside the array.

**Review by allhandsondeck — CHANGES_REQUESTED** [review]
Good job Nojdar!

File: homework/hackyourtemperature/_tests_/app.test.js
Comment (allhandsondeck) [inline homework/hackyourtemperature/_tests_/app.test.js:1]: Phrasing pattern followed here is nice in terms of [BDD](https://medium.com/javascript-scene/behavior-driven-development-bdd-and-functional-testing-62084ad7f1f2)(Behaviour Driven Development).
One improvement can be that to specifying the HTTP methods as well.

Please have a look at the example in MAKEME.md for that:
```js
describe("POST /", () => {
  it("Quick test", () => {
    expect(1).toBe(1);
  });
});
```

How can we rephrase or regroup these scenario so we can both understand which HTTP method or call it this scenario related to and also we can read the scenarios like you have written at the moment?

**Review by allhandsondeck — CHANGES_REQUESTED** [review]

File: homework/hackyourtemperature/package.json [hunk 8]
```diff
@@ -0,0 +1,30 @@
+{
+  "name": "hackyourtemperature",
+  "version": "1.0.0",
+  "description": "",
+  "type": "module",
+  "main": "server.js",
+  "scripts": {
+    "test": "jest",
+    "start": "node server.js",
+    "dev": "nodemon server.js"
```
Comment (allhandsondeck) [inline homework/hackyourtemperature/package.json:10]: Nice!

File: homework/hackyourtemperature/package.json [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
+  "dependencies": {
+    "express": "^4.18.2",
+    "express-handlebars": "^7.0.1",
+    "http": "0.0.1-security",
```
Comment (allhandsondeck) [inline homework/hackyourtemperature/package.json:18]: Is this `http` package in use?

File: homework/hackyourtemperature/app.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    );
+    const jsonData = await data.json();
+    const temperature = jsonData.main.temp;
+    res.send({ city: cityName, temperature: temperature });
+  } catch {
+    res.status(400).send({ weatherText: "City is not found!" });
```
Comment (allhandsondeck) [inline homework/hackyourtemperature/app.js:24]: Please also re-consider this requirement:
``` markdown
4. If the result is found, we also send back the object. Only, instead of just a string `City is not found!` dynamically add in the `cityName` and temperature (gotten from the result of the API call). Hint: use template strings to add variables in your strings!
```

File: homework/hackyourtemperature/_tests_/app.test.js (hunk 4 above)
Comment (NojdarTaho) [reply homework/hackyourtemperature/_tests_/app.test.js:19]: Here, I changed stringContaining to toMatch

File: homework/hackyourtemperature/_tests_/app.test.js (hunk 6 above)
Comment (NojdarTaho) [reply homework/hackyourtemperature/_tests_/app.test.js:35]: 👍

File: homework/hackyourtemperature/jest.config.js (hunk 7 above)
Comment (NojdarTaho) [reply homework/hackyourtemperature/jest.config.js:7]: I think is has to do with using import fetch from 'node-fetch'


 import http from 'node:http';
    ^^^^^^

    SyntaxError: Cannot use import statement outside a module

      1 | import express from "express";
    > 2 | import fetch from "node-fetch";
        | ^
      3 | import keys from "./sources/keys.js";
      4 | const app = express();
      5 |

File: homework/hackyourtemperature/_tests_/app.test.js
Comment (NojdarTaho) [reply homework/hackyourtemperature/_tests_/app.test.js:1]: I added an outer describe that indicates the  I am testing for POST http method following the example you gave me:

describe("POST /", () => {
})

File: homework/hackyourtemperature/package.json (hunk 9 above)
Comment (NojdarTaho) [reply homework/hackyourtemperature/package.json:18]: I do not think it is. I removed it

File: homework/hackyourtemperature/app.js (hunk 10 above)
Comment (NojdarTaho) [reply homework/hackyourtemperature/app.js:24]: 👍

**Comment by NojdarTaho** [comment]
Hey Utku, thank you so much for the great feedbacks. I tried to implement them. I hope my code is better now
